### PR TITLE
Add DFA construction helpers

### PR DESCRIPTION
**Pull request by Tagl**
Added the following construction helpers:
- `universal_language`
- `empty_language`
- `from_substring`
- `from_subsequence`
- `from_prefix`
- `from_suffix`
- `of_length`
- `count_mod`
- `nth_from_end`
- `nth_from_start`

### CODE REVIEW

**Comment by eliotwrobson** [comment]
@Tagl I actually have the `contains_substring` function written, along with some extra tests you might find useful for the `contains_subsequence`. Would you like me to post them here so you can include in this PR?

File: tests/test_dfa.py
Comment (eliotwrobson) [inline tests/test_dfa.py:1508]: Higher up in this file, there's a test that uses DeMorgan's laws along with the empty and universal languages. It would be good to use the functions you wrote in those tests. I think there may also be other tests that use the empty language (and include redundant generator code for those DFAs).

**Comment by Tagl** [comment]
> @Tagl I actually have the `contains_substring` function written, along with some extra tests you might find useful for the `contains_subsequence`. Would you like me to post them here so you can include in this PR?

Absolutely, let's try to get most of these useful construction functions together into this one PR

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:662]: Doesn't this only work if there are two input symbols?

File: tests/test_dfa.py
Comment (Tagl) [reply tests/test_dfa.py:1508]: Yeah I'm thinking a lot of tests can be updated since many of the DFA we test with are either "contains this pattern" or "avoids this pattern" where the patterns are either substrings or subsequences.

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:628]: I think it's better to use `None` as a sentinel value, just because using a float means passing in two different types into this function. @caleb531 what do you think?

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:649]: Kindof a nitpick, but using `~` actually involves some unnecessary function call overhead (since this operation minimizes by default), I think it would be better just to hardcode this construction (since it's only a couple lines long and can be copied from elsewhere).

**Comment by eliotwrobson** [comment]
@Tagl Per discussion, here's the substring and subsequence creation functions I had:
```python
    @classmethod
    def contains_subsequence(cls, subsequence, input_symbols):
        """
        Directly computes the minimal DFA recognizing strings containing the
        given subsequence.
        """
        transitions = {
            i: {
                symbol: i+1 if symbol == chr else i
                for symbol in input_symbols
            }
            for i, chr in enumerate(subsequence)
        }

        final_state = len(subsequence)
        transitions[final_state] = {
            symbol: final_state for symbol in input_symbols
        }

        return cls(
            states=set(transitions.keys()),
            input_symbols=input_symbols,
            transitions=transitions,
            initial_state=0,
            final_states={final_state},
        )

    @classmethod
    def contains_substring(cls, substring, input_symbols):
        """
        Directly computes the minimal DFA recognizing strings containing the
        given substring.
        """
        prefixes = [substring[:i] for i in range(len(substring))]

        transitions = {prefix: dict() for prefix in prefixes}
        transitions[substring] = {
            symbol: substring for symbol in input_symbols
        }

        for prefix in prefixes:
            prefix_dict = transitions.setdefault(prefix, dict())

            for symbol in input_symbols:
                # Look for next state after reading in the given input symbol
                possible_suffix = prefix + symbol

                # This while loop will always terminate, since the empty string is in the dict
                while possible_suffix not in transitions:
                    possible_suffix = possible_suffix[1:]

                prefix_dict[symbol] = possible_suffix

        return cls(
            states=set(transitions.keys()),
            input_symbols=input_symbols,
            transitions=transitions,
            initial_state='',
            final_states={substring},
        )
```
And here are the tests I mentioned
```python
def test_contains_substring(self):
        """Should compute the minimal DFA accepting strings with the given substring"""

        input_symbols = {'a', 'n', 'o', 'b'}

        equiv_dfa = DFA(
            states={'', 'n', 'na', 'nan', 'nano'},
            input_symbols=input_symbols,
            transitions={
                '': {'a': '', 'n': 'n', 'o': '', 'b': ''},
                'n': {'a': 'na', 'n': 'n', 'o': '', 'b': ''},
                'na': {'a': '', 'n': 'nan', 'o': '', 'b': ''},
                'nan': {'a': 'na', 'n': 'n', 'o': 'nano', 'b': ''},
                'nano': {'a': 'nano', 'n': 'nano', 'o': 'nano', 'b': 'nano'},
            },
            initial_state='',
            final_states={'nano'}
        )

        substring_dfa = DFA.contains_substring('nano', input_symbols)
        self.assertEqual(len(substring_dfa.states), len(substring_dfa.minify().states))

        self.assertEqual(len(substring_dfa.states), len(equiv_dfa.states))
        self.assertEqual(substring_dfa, equiv_dfa)

        subset_dfa = DFA.from_finite_language(['nano', 'bananano', 'nananano', 'naonano'], input_symbols)
        self.assertTrue(subset_dfa < substring_dfa)

    def test_contains_subsequence(self):
        """Should compute the minimal DFA accepting strings with the given subsequence"""

        input_symbols = {'a', 'n', 'o', 'b'}

        equiv_dfa = DFA(
            states={'', 'n', 'na', 'nan', 'nano'},
            input_symbols=input_symbols,
            transitions={
                '': {'a': '', 'n': 'n', 'o': '', 'b': ''},
                'n': {'a': 'na', 'n': 'n', 'o': 'n', 'b': 'n'},
                'na': {'a': 'na', 'n': 'nan', 'o': 'na', 'b': 'na'},
                'nan': {'a': 'nan', 'n': 'nan', 'o': 'nano', 'b': 'nan'},
                'nano': {'a': 'nano', 'n': 'nano', 'o': 'nano', 'b': 'nano'},
            },
            initial_state='',
            final_states={'nano'}
        )

        subsequence_dfa = DFA.contains_subsequence('nano', input_symbols)
        self.assertEqual(len(subsequence_dfa.states), len(subsequence_dfa.minify().states))

        self.assertEqual(len(subsequence_dfa.states), len(equiv_dfa.states))
        self.assertEqual(subsequence_dfa, equiv_dfa)

        subset_dfa = DFA.from_finite_language(['naooono', 'bananano', 'onbaonbo', 'ooonano'], input_symbols)
        self.assertTrue(subset_dfa < subsequence_dfa)

        substring_dfa = DFA.contains_substring('nano', input_symbols)
        self.assertTrue(substring_dfa < subsequence_dfa)
```

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py:662]: You have the symbol `symbol` which is the one you are looking for to be the start of your $n$ character suffix.
All other symbols belong to the same equivalence class of *not what you're looking for*. Therefore the construction looks something like this:
image

Here we are looking for `x` and can treat all other characters as `?`

Comment (eliotwrobson) [reply automata/fa/dfa.py:662]: Ahhhh I see now, thanks for the clarification 😃

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py:628]: On that note, I think we should make input_symbols the first parameter in these construction methods so we can make easier use of default parameters.

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: Updated

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py:708]: We might want to leave a comment in the code regarding this as it is not immediately obvious.

File: tests/test_dfa.py
Comment (Tagl) [reply tests/test_dfa.py:1461]: I've updated some of the old tests. I didn't want to change anything to do with equality since in those it's best to have some manual constructions.

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py:662]: Updated the ordering so now `input_symbols` is the first parameter

**Comment by eliotwrobson** [comment]
@Tagl Containing multiple copies of a substring might be interesting? Also possibly counting the length / number of symbols mod k (where k is an input parameter). My ideas are starting to get a little contrived. Though the new functions so far look awesome.

File: automata/fa/dfa.py
Comment (caleb531) [inline automata/fa/dfa.py:747]: @Tagl Per my comment above, can we rename this method to `from_subsequence`?

File: automata/fa/dfa.py
Comment (caleb531) [inline automata/fa/dfa.py:713]: @Tagl Can we rename this method to `from_substring`? That would map better nicely to the Python convention used in this library (e.g. `from_dfa`, `from_nfa`, etc.)

File: automata/fa/dfa.py
Comment (caleb531) [inline automata/fa/dfa.py:767]: Wondering if keyword-only args would be good to use here, at least for `min_length` and `max_length. I think it would make your README example a bit cleaner:

```python
def of_length(cls, input_symbols, *, min_length=0, max_length=float('inf')):
    ...
```

```python
dfa = DFA.of_length({'0', '1'}, min_length=4, max_length=float('inf'))
```

@eliotwrobson, what do you think?

File: automata/fa/dfa.py
Comment (caleb531) [reply automata/fa/dfa.py:844]: @Tagl Agreed about adding a code comment (my future self will probably find it helpful).

File: README.md
Comment (caleb531) [inline README.md:399]: @Tagl Please add a note to the documentation here about the `ValueError` that's raised if the `symbol` does not exist within `input_symbols`.

File: tests/test_dfa.py
Comment (caleb531) [inline tests/test_dfa.py:687]: @Tagl Similar with the past PR, please use lowercase variable names instead of `A`, `B`, `C`, etc.

**Review by caleb531 — CHANGES_REQUESTED** [review]
@Tagl Looks good so far—left a few requests for you to review.

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:767]: Sounds good, in keeping with the conventions of other functions. I think this got mentioned before, but I think that `None` is a better default sentinel value than `float('inf')`.

File: README.md
Comment (eliotwrobson) [reply README.md:399]: Isn't there a specialized `InvaildSymbol` exception somewhere?

Comment (caleb531) [reply README.md:399]: @eliotwrobson @Tagl Ah yes, `InvalidSymbolError`. Good call.

**Comment by Tagl** [comment]
@eliotwrobson @caleb531 `from_prefix` and `from_suffix` are probably quite commonly useful too.

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py:767]: We should be able to pass whatever is returned by the word length functions in to this function, in my opinion. The default sentinel value can be None but it must also support float('inf') if the `maximum_word_length` function returns float('inf').
If we switch to None then we still are passing in two different types.

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py:713]: I can do that and with that naming change I would like to propose an additional optional parameter which defines whether we seek containment or avoidance.
Construction remains the same but final states will be different. This allows the user to avoid the overhead of using the complement function after construction.
My main use case for example has a lot to do with containment/avoidance of subsequences.

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py:849]: Comment added

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: Updated to keyword arguments

File: README.md
Comment (Tagl) [reply README.md:399]: Updated

File: tests/test_dfa.py
Comment (Tagl) [reply tests/test_dfa.py]: Everything I touched in this PR has been updated

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: Updated, see further discussion on other comment

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py]: Right, but I think `None` is known to be a sentinel value (enough that Python has native support for it with the `Optional` type), while the `float('inf')` is requiring the user to passing _one specific value_ of the `float` type (they can't put in `3.5`, for example). I think it's actually worth considering whether we should remove all of the `float('inf')` return types altogether for this. Ultimately it's up to @caleb531, since is more about how the library is organized rather than any serious algorithmic considerations.

Comment (caleb531) [reply automata/fa/dfa.py]: I would agree with @eliotwrobson that `None` is preferable. It is a Python convention, and semantically, it implies that "there is no maximum length" rather than "the maximum length is infinity" (the former of which makes more sense to me).

Plus, it is unnatural to me to represent a length value with a float, per @eliotwrobson's other point. Sure, we could add a library to represent infinity as an integer, but that seems a bit overkill to me, and erodes my first argument above.

Comment (Tagl) [reply automata/fa/dfa.py]: I'm fine with using `None`, there's just pros and cons with each approach.
`None` also semantically makes sense for the empty language, but there it means something completely different.
I think leaving that case as an exception though makes more sense since it truly is an exception.

I don't really like using floating points at all when I can avoid it. The reason I chose it originally was simplicity, you can compare using the `<` operator and everything will work the same without special casing.

I think we should also avoid adding dependencies unless it is heavily relied upon or adds great value. If we wanted to go that way, creating an integer data type with a sentinel infinity value is relatively simple and does not warrant adding a dependency in my opinion.

I will change all the logic to use `None` then.

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py]: @Tagl I think that parameter would be great. We can convert it to an internal helper if later on we feel like we want different user-facing function names.

**Comment by caleb531** [comment]
@Tagl Just a heads up—I was refactoring the assertions in `test_dfa` and it seems to have caused some conflicts on this PR. You'll probably want to rebase on top of the latest `develop`—let me know if I can help with that.

Apologies for the inconvenience.

**Comment by Tagl** [comment]
> @Tagl Just a heads up—I was refactoring the assertions in `test_dfa` and it seems to have caused some conflicts on this PR. You'll probably want to rebase on top of the latest `develop`—let me know if I can help with that.
> 
> Apologies for the inconvenience.

But I already refactored those tests 😢

Which version do you prefer? I find naming the automata something other than `dfa` when we are working with multiple together is easier to read.

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: > On a related note, do you feel like having a Match object would be useful to you? Something that takes in an FSM (DFA or NFA) and uses it to search for all matches in a given input string.

No, not particularly useful for me so far.

**Comment by caleb531** [comment]
@Tagl I'm sorry... I will make it up to you by taking care of the rebase.

That refactor I made was mostly to refactor the uppercase single-letter names like `A` and `B. I hear what you're saying, and I have no qualms with instance names like `at_least_four_ones`.

**Comment by caleb531** [comment]
@Tagl I got frustrated with the rebase, so I decided to revert my test_dfa commit and not rebase your branch at all. I'll tweak test_dfa after this branch gets merged. 🙂

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: Now my only concern is the `cardinality` function. I feel like `None` is not the proper value to return so I'm inclined to have it raise a `ValueError`, just as `len` must do.

Comment (Tagl) [reply automata/fa/dfa.py]: Updated

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: @eliotwrobson @caleb531 I just pushed an optimization of this function. I was able to pass in strings of length `>100000` after the optimization with a much faster runtime than before.

Example repeated pattern `abcd`
| size of string | runtime in seconds |
| ----:| ----------------------:|
| 4 | 0.00027 |
| 40 | 0.00023 |
| 400 | 0.00234 |
| 4000 | 0.02422 |
| 40000 | 0.19565 |
| 400000 | 1.93628 |

**Comment by Tagl** [comment]
> @Tagl Containing multiple copies of a substring might be interesting? Also possibly counting the length / number of symbols mod k (where k is an input parameter). My ideas are starting to get a little contrived. Though the new functions so far look awesome.

Length or symbol count modulo $k$ could be useful so I will add that but I think that will be the last thing I put in this PR and then I'll update the documentation.

**Comment by Tagl** [comment]
@caleb531 @eliotwrobson Should be all ready now for review.

File: README.md
Comment (caleb531) [inline README.md:273]: @Tagl Didn't even realize the hierarchy was incorrect. Thanks for fixing this. 🙂

Comment (Tagl) [reply README.md:273]: You're most welcome

**Review by caleb531 — APPROVED** [review]
@Tagl Looks great! Thank you for this.

**Review by eliotwrobson — APPROVED** [review]
Looks really awesome, thank you for all the hard work @Tagl !

**Pull request merged**
